BoxContentSDK/BoxContentSDK/OAuth2/BOXAuthorizationViewController.m

```diff
@@ -241,38 +250,52 @@ - (void)failAuthenticationWithConnection:(NSURLConnection *)connection
     self.connection = nil;
     self.connectionResponse = nil;
     [self setWebViewCanBeUsedDirectly:NO forHost:connection.currentRequest.URL.host];
-
+    
+    self.connectionError = error;
+    self.connectionErrorMessage = message;
+    
+    // We can only handle the connection error if we're in the view hierarchy. It would be difficult to display UIAlertControllers when not sure we're in the view hierarchy. We will try again after viewDidAppear
     if (self.view.window) {
-        
-        self.connectionError = error;
-        
-        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Unable to Log In", @"Alert view title: Title for failed SSO login due to authentication issue")
-                                                                                 message:message
-                                                                          preferredStyle:UIAlertControllerStyleAlert];
-        
-        __weak typeof(self) weakSelf = self;
-        void (^completion)(UIAlertAction *action) = ^void(UIAlertAction *action) {
-            [alertController dismissViewControllerAnimated:YES completion:nil];
-            
-            if (weakSelf.completionBlock) {
-                NSError *error = weakSelf.connectionError;
-                weakSelf.connectionError = nil;
-                weakSelf.completionBlock(weakSelf, nil, error);
-            }
-            // Clear out the authentication challenge in memory
-            weakSelf.authenticationChallenge = nil;
-        };
-        
-        UIAlertAction *okAction = [UIAlertAction actionWithTitle:NSLocalizedString(@"OK", @"Label: Allow the user to accept the current condition, often used on buttons to dismiss alerts")
-                                                           style:UIAlertActionStyleDefault
-                                                         handler:completion];
-        
-        [alertController addAction:okAction];
-        [self presentViewController:alertController animated:YES completion:nil];
+        [self handleConnectionError];
+        // Nil out to avoid any view did appear extra call.
+        self.connectionError = nil;
+        self.connectionErrorMessage = nil;
+    } else {
+        // We still want to report the completion block so that the necessary cleanup is done up in the call tree.
+        if (self.completionBlock) {
+            self.completionBlock(self, nil, self.connectionError);
```

Comment:
After thinking about it, we need it here. We have several cases we need to handle, and we can't only handle the error in one place without missing one of them : 
- Error when the controller is not presented (as done in flexo, viewDidAppear will never get called, need to call the completion block)
- Error when presented but viewDidAppear was not called yet : we handle the error and then make sure it won't be handled by viewDidAppear
- Error after viewDidAppear is called : handle the error.